implementations/rust/ockam/ockam_identity/src/purpose_keys/builder/credential.rs

```diff
@@ -0,0 +1,121 @@
+use ockam_core::compat::sync::Arc;
+use ockam_core::Result;
+use ockam_vault::{SigningKeyType, SigningSecretKeyHandle};
+
+use crate::models::{PurposePublicKey, TimestampInSeconds};
+use crate::purpose_keys::Ttl;
+use crate::{CredentialPurposeKey, Identifier, Purpose, PurposeKeyCreation};
+
+/// Default TTL for an Identity key
+pub const DEFAULT_CREDENTIAL_PURPOSE_KEY_TTL: TimestampInSeconds =
+    TimestampInSeconds(5 * 365 * 24 * 60 * 60); // Five years
+
+enum Key {
+    Generate(SigningKeyType),
+    Existing(SigningSecretKeyHandle),
+}
+
+/// Builder for [`CredentialPurposeKey`]
+pub struct CredentialPurposeKeyBuilder {
+    purpose_keys_creation: Arc<PurposeKeyCreation>,
+
+    identifier: Identifier,
+    key: Key,
+    ttl: Ttl,
+}
+
+impl CredentialPurposeKeyBuilder {
```

Comment:
What you suggest is exactly what I attempted initially (and succeeded). But because type-safety increased the separation between two types of PurposeKeys, it caused a lot of code duplications between the types which were hard to reuse (this version was bigger by around 300-400 lines of code). I decided that the way it is now is better, with its cons.